### PR TITLE
fix: allow parsing of empty keys in objects

### LINKED ISSUE #432: qs.parse omitting empty string keys
Hi, it will be nice to have this part configurable:

https://github.com/ljharb/qs/blob/542a5c7ff88d7229efa2e22c7c8a7d69375f5e72/lib/parse.js#L150-L152

current output:
```ts
qs.parse("=1&=2") === {}
qs.parse(" =1& =2") === { " ": ["1","2"] }
```

expected output :
```ts
qs.parse("=1&=2", { allowEmptyKeys: true }) === { "": ["1","2"] }
```

### PR DESCRIPTION
**Pull request by Coobaha**
Resolves #432 

- [x] to add an option to both parse and stringify
- [ ] to have tests, for both parse and stringify, for at least:
  - [x] repeat keys with all array formats
  - [x] bracketed keys with all array formats
  - [x] a lone = (and =& and &= and &=&)
  - [ ] ❌ deep array/object case  

**If we are happy with this implementation and want to proceed:**

 - [x]  to document in the readme
 - [x]  split tests before merge to parse/stringify files
 - [ ] update https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/qs (also `allowSparse` is missing there and maybe other types)
 
 
 Other:
 - Stringify works without an option, so it is potentially a patch but altering parse behaviour only via option seems a safe bet to not accidentally break ppl code, and can be considered as a default behaviour in the next major version?

### CODE REVIEW

File: test/parse.js [hunk 1]
```diff
@@ -843,3 +843,52 @@ test('parse()', function (t) {
...
+        st.deepEqual(qs.stringify({ '': ['a', 'b'] }), '%5B0%5D=a&%5B1%5D=b');
+        st.deepEqual(qs.stringify({ '': ['a', 'b'], a: [1, 2] }, { encode: false }), '[0]=a&[1]=b&a[0]=1&a[1]=2');
+
+        st.end();
+    });
+    t.test('edge case with object/arrays', function (st) {
+
+        st.deepEqual(qs.stringify({ '': { '': [2, 3] } }, { encode: false }), '[][0]=2&[][1]=3');
+
+        st.deepEqual(qs.parse('[][0]=2&[][1]=3', { allowEmptyKeys: true }), { '': { '': [2, 3] } });
```
Comment (Coobaha) [inline test/parse.js:890]: This fails case fails, but this is ambiguous string which is treated as `{ : [ '2', '3' ] }` and stringify <-> parse is not producing the same results

File: lib/parse.js [hunk 2]
```diff
@@ -83,6 +84,9 @@ var parseValues = function parseQueryStringValues(str, options) {
         var key, val;
         if (pos === -1) {
             key = options.decoder(part, defaults.decoder, charset, 'key');
+            if (key === '') {
+                continue;
```
Comment (Coobaha) [inline lib/parse.js:88]: handles lonely `&`

File: test/parse.js [hunk 3]
```diff
@@ -843,3 +843,53 @@ test('parse()', function (t) {
...
+        st.deepEqual(qs.stringify({ '': ['a', 'b'], a: [1, 2] }, { encode: false }), '[0]=a&[1]=b&a[0]=1&a[1]=2');
+
+        st.end();
+    });
+    t.test('edge case with object/arrays', function (st) {
+
+        st.deepEqual(qs.stringify({ '': { '': [2, 3] } }, { encode: false }), '[][0]=2&[][1]=3');
+        st.deepEqual(qs.stringify({ '': { '': [2, 3], a: 2 } }, { encode: false }), '[][0]=2&[][1]=3&[a]=2');
+        st.deepEqual(qs.parse('[][0]=2&[][1]=3&[a]=2', { allowEmptyKeys: true }), { '': { '': ['2', '3'] } });
+        st.deepEqual(qs.parse('[][0]=2&[][1]=3&[a]=2', { allowEmptyKeys: true }), { '': { '': ['2', '3'], a: '2' } });
```
Comment (Coobaha) [inline test/parse.js:891]:
```tape
  ---
    operator: deepEqual
    expected: |-
      { : { : [ '2', '3' ] } }
    actual: |-
      { : { 0: '2', 1: '3', a: '2' } }
```

```
    operator: deepEqual
    expected: |-
      { : { : [ '2', '3' ], a: '2' } }
    actual: |-
      { : { 0: '2', 1: '3', a: '2' } }
```

File: test/parse.js [hunk 4]
```diff
@@ -843,3 +843,53 @@ test('parse()', function (t) {
...
+        st.deepEqual(qs.stringify({ '': ['a', 'b'], a: [1, 2] }, { encode: false }), '[0]=a&[1]=b&a[0]=1&a[1]=2');
+
+        st.end();
+    });
+    t.test('edge case with object/arrays', function (st) {
+
+        st.deepEqual(qs.stringify({ '': { '': [2, 3] } }, { encode: false }), '[][0]=2&[][1]=3');
+        st.deepEqual(qs.stringify({ '': { '': [2, 3], a: 2 } }, { encode: false }), '[][0]=2&[][1]=3&[a]=2');
+        st.deepEqual(qs.parse('[][0]=2&[][1]=3', { allowEmptyKeys: true }), { '': { '': ['2', '3'] } });
+        st.deepEqual(qs.parse('[][0]=2&[][1]=3&[a]=2', { allowEmptyKeys: true }), { '': { '': ['2', '3'], a: '2' } });
```
Comment (Coobaha) [inline test/parse.js:891]: This fails case fails, but this is ambiguous string which is treated as { : [ '2', '3' ] } and stringify <-> parse is not producing the same results

```
  ---
    operator: deepEqual
    expected: |-
      { : { : [ '2', '3' ] } }
    actual: |-
     { : [ '2', '3' ] }

```

```
    operator: deepEqual
    expected: |-
      { : { : [ '2', '3' ], a: '2' } }
    actual: |-
      { : { 0: '2', 1: '3', a: '2' } }
```

**Comment by ljharb** [comment]
There's unlikely to be a major version any time soon; breaking changes just aren't worth it. Configuring it via option is the best bet for now.

File: lib/parse.js (hunk 2 above)
Comment (ljharb) [reply lib/parse.js:88]: what if the key already exists? this would bypass the `utils.combine` logic below

File: lib/parse.js [hunk 5]
```diff
@@ -148,7 +152,7 @@ var parseObject = function (chain, val, options, valuesParsed) {
 };
 
 var parseKeys = function parseQueryStringKeys(givenKey, val, options, valuesParsed) {
-    if (!givenKey) {
+    if (options.allowEmptyKeys ? givenKey !== '' && !givenKey : !givenKey) {
```
Comment (ljharb) [inline lib/parse.js:155]:
```suggestion
    if (!givenKey && (!options.allowEmptyKeys || givenKey !== '')) {
```
to avoid the inline ternary

File: test/parse.js [hunk 6]
```diff
@@ -843,3 +843,53 @@ test('parse()', function (t) {
 
     t.end();
 });
+
+test('parse/stringify empty key WIP', function (t) {
+    t.test('parses an object with empty string key', function (st) {
+        st.deepEqual(qs.parse('=a', { allowEmptyKeys: true }), { '': 'a' });
```
Comment (ljharb) [inline test/parse.js:849]: for each of these test inputs, let's test both allowEmptyKeys true, false, and absent

File: test/parse.js (hunk 4 above)
Comment (ljharb) [reply test/parse.js:891]: it's ok to mark this case as a TODO for now; there's an existing issue about a nested array/object combo

**Review by ljharb — COMMENTED** [review]
At the very least, let's add stringify tests using all of the outputs in the parse tests as input.

File: test/parse.js (hunk 6 above)
Comment (Coobaha) [reply test/parse.js]: Done, please let me know what will be the correct place for this tests? They are testing both parse and stringify and current tests are split by this methods. To me it feels more reasonable to test them together in this case.

File: lib/parse.js (hunk 2 above)
Comment (Coobaha) [reply lib/parse.js:88]: Could you provide input for this case? All previous inputs are still valid and i've covered this cases in tests:
```
&
&=
&=&
&=&=
&=&=&
=
=&
=&&&
=&=&=&
``` 

Do i miss something? Thanks

Comment (ljharb) [reply lib/parse.js:88]: I don't have any input off the top of my head :-) just looking at the code paths.

File: test/parse.js [hunk 7]
```diff
@@ -843,3 +843,70 @@ test('parse()', function (t) {
...
+        { input: '%5B0%5D=a&%5B1%5D=b', withEmptyKeys: { '': ['a', 'b'] }, stringifyOutput: '[0]=a&[1]=b', noEmptyKeys: { 0: 'a', 1: 'b' } }
+    ];
+
+    testCases.forEach(function (testCase) {
+        t.test('parses an object with empty string key with ' + testCase.input, function (st) {
+            var parsed = qs.parse(testCase.input, { allowEmptyKeys: true });
+
+            st.deepEqual(parsed, testCase.withEmptyKeys);
+            st.deepEqual(qs.stringify(parsed, { encode: false }), testCase.stringifyOutput);
+            st.deepEqual(qs.parse(testCase.stringifyOutput, { allowEmptyKeys: true }), testCase.withEmptyKeys);
```
Comment (ljharb) [inline test/parse.js:888]: i'd prefer to keep stringify tests in the stringify tests file; it's fine if the test cases are in a separate file both test files require.

Comment (Coobaha) [reply test/parse.js]: Hey, sorry for the delay :) i've extracted test cases into own file and and separated stringify/parse parts accordingly

File: test/parse.js [hunk 8]
```diff
@@ -843,3 +844,46 @@ test('parse()', function (t) {
...
+            { '': { '': ['2', '3'] } },
+            'array/object conversion',
+            { skip: 'TODO: figure out what this should do' }
+        );
+
+        st.deepEqual(
+            qs.parse('[][0]=2&[][1]=3&[a]=2', { allowEmptyKeys: true }),
+            { '': { '': ['2', '3'], a: '2' } },
+            'array/object conversion',
+            { skip: 'TODO: figure out what this should do' }
```
Comment (ljharb) [inline test/parse.js:884]: let's discuss these TODOs. what are your thoughts?

File: test/stringify.js [hunk 9]
```diff
@@ -861,3 +862,20 @@ test('stringify()', function (t) {
 
     t.end();
 });
+
+test('stringifies empty keys', function (t) {
+    emptyTestCases.forEach(function (testCase) {
+        t.test('stringifies an object with empty string key with ' + testCase.input, function (st) {
+            st.deepEqual(qs.stringify(testCase.withEmptyKeys, { encode: false }), testCase.stringifyOutput);
```
Comment (ljharb) [inline test/stringify.js:869]: can we also add a test case where the input is a stringified `noEmptyKeys`?

**Review by ljharb — COMMENTED** [review]
Looking great!

File: test/parse.js (hunk 8 above)
Comment (Coobaha) [reply test/parse.js:884]: Since we parse empty keys I assume output should be equal to key-full strings? 
- If we use `a[b][0]=2&a[b][1]=3` the output object is `{ a: { b: [ '2', '3' ] } }`
- when we replace `a` and `b` with empty strings in this object: `{ '': { '': [ '2', '3' ] } }` and query string should be `[][0]=2&[][1]=3` which is `<EMPTY_KEY>[<EMPTY_KEY>][0]=2` 

Current output for `[][0]=2&[][1]=3` without empty keys flag is ` { 0: '2', 1: '3' }`

File: test/empty-keys-cases.js [hunk 10]
```diff
@@ -2,36 +2,36 @@
...
-        { input: '&=', withEmptyKeys: { '': '' }, stringifyOutput: '=', noEmptyKeys: {} },
-        { input: '&=&', withEmptyKeys: { '': '' }, stringifyOutput: '=', noEmptyKeys: {} },
-        { input: '&=&=', withEmptyKeys: { '': ['', ''] }, stringifyOutput: '[0]=&[1]=', noEmptyKeys: {} },
-        { input: '&=&=&', withEmptyKeys: { '': ['', ''] }, stringifyOutput: '[0]=&[1]=', noEmptyKeys: {} },
+        { input: '&', withEmptyKeys: {}, stringifyOutput: '', noEmptyKeys: {}, stringifyOutputNoEmpty: '' },
+        { input: '&&', withEmptyKeys: {}, stringifyOutput: '', noEmptyKeys: {}, stringifyOutputNoEmpty: '' },
+        { input: '&=', withEmptyKeys: { '': '' }, stringifyOutput: '=', noEmptyKeys: {}, stringifyOutputNoEmpty: '' },
+        { input: '&=&', withEmptyKeys: { '': '' }, stringifyOutput: '=', noEmptyKeys: {}, stringifyOutputNoEmpty: '' },
+        { input: '&=&=', withEmptyKeys: { '': ['', ''] }, stringifyOutput: '[0]=&[1]=', noEmptyKeys: {}, stringifyOutputNoEmpty: '' },
+        { input: '&=&=&', withEmptyKeys: { '': ['', ''] }, stringifyOutput: '[0]=&[1]=', noEmptyKeys: {}, stringifyOutputNoEmpty: '' },
```
Comment (ljharb) [inline test/empty-keys-cases.js:10]: it makes me *very* nervous to be modifying test cases for a bug fix. is there any chance for a default algorithm to infer `stringifyOutputNoEmpty` rather than needing to add it to each case?

Comment (Coobaha) [reply test/empty-keys-cases.js:10]: 🤷 I think explicit cases are better here. I assume you thought that we should default to `''` if you want a smaller diff, then it also makes to do the same for `stringifyOutput`?
